### PR TITLE
[fix](memory) Fix LRU cache stale sweep

### PR DESCRIPTION
**Pull request by xinyiZzz**
## Proposed changes

Remove `LRUCacheValueBase`, put `last_visit_time` into `LRUHandle`, and automatically update timestamp to `last_visit_time` during cache insert and lookup.

Do not rely on external modification of `last_visit_time`, which is often forgotten.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/olap/lru_cache.cpp [hunk 1]
```diff
@@ -622,18 +635,18 @@ uint64_t ShardedLRUCache::new_id() {
...
     int64_t num_prune = 0;
     for (int s = 0; s < _num_shards; s++) {
-        num_prune += _shards[s]->prune();
+        num_prune += _shards[s]->prune(freed_size);
     }
     return num_prune;
 }
 
-int64_t ShardedLRUCache::prune_if(CacheValuePredicate pred, bool lazy_mode) {
+int64_t ShardedLRUCache::prune_if(CachePrunePredicate pred, int64_t* freed_size, bool lazy_mode) {
```
Comment (yiguolei) [inline be/src/olap/lru_cache.cpp:646]: 一般把返回值，放到参数列表的最后面

File: be/src/olap/lru_cache.cpp [hunk 2]
```diff
@@ -458,23 +461,28 @@ int64_t LRUCache::prune() {
...
         to_remove_head = next;
     }
+    if (freed_size) {
+        *freed_size = size;
+    }
     return pruned_count;
 }
 
-int64_t LRUCache::prune_if(CacheValuePredicate pred, bool lazy_mode) {
+int64_t LRUCache::prune_if(CachePrunePredicate pred, int64_t* freed_size, bool lazy_mode) {
```
Comment (yiguolei) [inline be/src/olap/lru_cache.cpp:478]: 一般把返回值放到参数列表的最后面

File: be/src/olap/lru_cache.h [hunk 3]
```diff
@@ -479,8 +489,11 @@ class DummyLRUCache : public Cache {
     void* value(Handle* handle) override;
     Slice value_slice(Handle* handle) override;
     uint64_t new_id() override { return 0; };
-    int64_t prune() override { return 0; };
-    int64_t prune_if(CacheValuePredicate pred, bool lazy_mode = false) override { return 0; };
+    int64_t prune(int64_t* freed_size = nullptr) override { return 0; };
+    int64_t prune_if(CachePrunePredicate pred, int64_t* freed_size = nullptr,
```
Comment (yiguolei) [inline be/src/olap/lru_cache.h:493]: 我觉得啊，你要不干脆返回std::tuple<free_count, free_size> 要不看着怪怪的，一个是return 返回值，一个是通过参数指针返回

File: be/src/olap/lru_cache.cpp (hunk 1 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:646]: 有道理，我是按参数使用频率排的…

File: be/src/olap/lru_cache.h (hunk 3 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.h:493]: done

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

File: be/src/olap/lru_cache.h [hunk 4]
```diff
@@ -246,15 +252,19 @@ struct LRUHandle {
...
+    size_t bytes;  // Used by LRUCacheType::NUMBER, LRUCacheType::SIZE equal to total_size.
+    bool in_cache; // Whether entry is in the cache.
     uint32_t refs;
     uint32_t hash; // Hash of key(); used for fast sharding and comparisons
     CachePriority priority = CachePriority::NORMAL;
     MemTrackerLimiter* mem_tracker;
     LRUCacheType type;
+    // Save the last visit time of this cache entry.
+    // Use atomic because it may be modified by multi threads.
+    int64_t last_visit_time;
```
Comment (jacktengg) [inline be/src/olap/lru_cache.h:265]: atomic or not?

File: be/src/olap/lru_cache.h [hunk 5]
```diff
@@ -219,12 +222,15 @@ class Cache {
     // encouraged to override the default implementation.  A future release of
     // leveldb may change prune() to a pure abstract method.
     // return num of entries being pruned.
-    virtual int64_t prune() { return 0; }
+    virtual std::tuple<PrunedCount, PrunedSize> prune() { return {0, 0}; }
```
Comment (JackDrogon) [inline be/src/olap/lru_cache.h:225]: use struct not tuple

File: be/src/olap/lru_cache.h (hunk 4 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.h:265]: `last_visit_time` only be modified in `LRUCache::insert` and `LRUCache::lookup`, in the cache lock.

File: be/src/olap/lru_cache.cpp [hunk 6]
```diff
@@ -622,20 +629,27 @@ uint64_t ShardedLRUCache::new_id() {
     return _last_id.fetch_add(1, std::memory_order_relaxed);
 }
 
-int64_t ShardedLRUCache::prune() {
-    int64_t num_prune = 0;
+std::tuple<PrunedCount, PrunedSize> ShardedLRUCache::prune() {
```
Comment (JackDrogon) [inline be/src/olap/lru_cache.cpp:632]: use range_view algorithm replace for

File: be/src/olap/lru_cache.cpp [hunk 7]
```diff
@@ -622,20 +629,27 @@ uint64_t ShardedLRUCache::new_id() {
...
+    return {pruned_count, pruned_size};
 }
 
-int64_t ShardedLRUCache::prune_if(CacheValuePredicate pred, bool lazy_mode) {
-    int64_t num_prune = 0;
+std::tuple<PrunedCount, PrunedSize> ShardedLRUCache::prune_if(CachePrunePredicate pred,
+                                                              bool lazy_mode) {
+    int64_t pruned_count = 0;
+    int64_t pruned_size = 0;
     for (int s = 0; s < _num_shards; s++) {
```
Comment (JackDrogon) [inline be/src/olap/lru_cache.cpp:647]: use range_view algorithm replace for

File: be/src/olap/lru_cache.h (hunk 4 above)
Comment (jacktengg) [reply be/src/olap/lru_cache.h:265]: Then better to remove the comment.

File: be/src/runtime/memory/lru_cache_policy.h [hunk 8]
```diff
@@ -84,26 +76,24 @@ class LRUCachePolicy : public CachePolicy {
 
     // Try to prune the cache if expired.
     void prune_stale() override {
+        COUNTER_SET(_freed_entrys_counter, (int64_t)0);
```
Comment (JackDrogon) [inline be/src/runtime/memory/lru_cache_policy.h:79]: use 0LL replace （int64_t) 0

File: be/src/olap/lru_cache.h (hunk 4 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.h:265]: Good! I forgot to change the comment.

**Review by JackDrogon — CHANGES_REQUESTED** [review]

File: be/src/runtime/memory/lru_cache_policy.h (hunk 8 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/lru_cache_policy.h:79]: 0L failed to compile on mac, let me try 0LL

Comment (xinyiZzz) [reply be/src/runtime/memory/lru_cache_policy.h:79]: warning: Call to member function 'set' is ambiguous

virtual void set(int64_t value)
virtual void set(double value)

File: be/src/olap/lru_cache.h (hunk 5 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.h:225]: done

File: be/src/olap/lru_cache.h (hunk 4 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.h]: done

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
run performance

**Review by yiguolei — APPROVED** [review]

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**